private/bufpkg/bufcas/manifest.go

```diff
@@ -0,0 +1,183 @@
+// Copyright 2020-2023 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bufcas
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"sort"
+	"strings"
+
+	storagev1beta1 "github.com/bufbuild/buf/private/gen/proto/go/buf/registry/storage/v1beta1"
+)
+
+// Maniest is a set FileNodes.
+type Manifest interface {
+	// fmt.Stringer encodes the Manifest into its canonical form, consisting of
+	// an ordered list of paths and their digests. Sorted by path.
+	//
+	// See the documentation on FileNode for how FileNodes are encoded.
+	//
+	// An example encoded manifest:
+	//
+	//	shake256:cd22db48cf7c274bbffcb5494a854000cd21b074df7c6edabbd0102c4be8d7623e3931560fcda7acfab286ae1d4f506911daa31f223ee159f59ffce0c7acbbaa  buf.lock
+	//	shake256:3b353aa5aacd11015e8577f16e2c4e7a242ce773d8e3a16806795bb94f76e601b0db9bf42d5e1907fda63303e1fa1c65f1c175ecc025a3ef29c3456ad237ad84  buf.md
+	//	shake256:7c88a20cf931702d042a4ddee3fde5de84814544411f1c62dbf435b1b81a12a8866a070baabcf8b5a0d31675af361ccb2d93ddada4cdcc11bab7ea3d8d7c4667  buf.yaml
+	//	shake256:9db25155eafd19b36882cff129daac575baa67ee44d1cb1fd3894342b28c72b83eb21aa595b806e9cb5344759bc8308200c5af98e4329aa83014dde99afa903a  pet/v1/pet.proto
+	fmt.Stringer
+
+	// FileNodes returns the set of FileNodes that make up the Manifest.
+	//
+	// The paths of the given FileNodes are guaranteed to be unique.
+	// The iteration order will be the sorted order of the paths.
+	FileNodes() []FileNode
+	// GetDigest gets the Digest for the given path.
+	//
+	// Returns nil if the path does not exist.
+	GetDigest(path string) Digest
+
+	// Protect against creation of a Manifest outside of this package, as we
+	// do very careful validation.
+	isManifest()
+}
+
+// NewManifest returns a new Manifest for the given path -> Digest map.
+//
+// FileNodes are deduplicated upon construction, however if two FileNodes
```

Comment:
It deduplicates the same as `BlobSet` does - it would seem OK to construct a `[]FileNode` via say walking a few buckets, and if there's a duplicate, filtering it out here. `BlobSet` we definitely want the deduplication`, here is more arguable, but we probably want to be consistent. Just my two cents.

Reply:
I changed `NewBlobSet` to return an `error` to give us more flexibility in the future to add validation without needing to refactor.